lib/Sandbox.js

```diff
@@ -204,35 +207,49 @@ function getEnvironment() {
 function buildDependencyTree(
   baseDir: string,
   dependencySpecList: Array<string>,
-  packageDependencyTrace: Array<string>
+  context: {
+    packageDependencyTrace: Array<string>;
+    packageCache: Map<string, PackageInfo>;
+  }
 ): DependencyTree {
   let dependencyTree: {[name: string]: PackageInfo} = {};
   for (let dependencySpec of dependencySpecList) {
     const {name} = parseDependencySpec(dependencySpec);
     const dependencyPackageJsonPath  = resolveSync(
       `${name}/package.json`, {basedir: baseDir});
-    const dependencyBaseDir = path.dirname(dependencyPackageJsonPath);
-    const packageJson = readPackageJson(dependencyPackageJsonPath);
-    const depSpecList = objectToDependencySpecList(
-      packageJson.dependencies,
-      packageJson.peerDependencies
-    );
 
-    dependencyTree[name] = {
-      version: packageJson.version,
-      source: packageJson._resolved || `local:${fs.realpathSync(dependencyBaseDir)}`,
-      sourceType: packageJson._resolved ? 'remote' : 'local',
-      rootDirectory: dependencyBaseDir,
-      packageJson,
-      normalizedName: normalizeName(packageJson.name),
-      dependencyTree: depSpecList.length > 0
-        ? buildDependencyTree(
-            dependencyBaseDir,
-            depSpecList,
-            packageDependencyTrace.concat(packageJson.name)
-          )
-        : {}
+    let packageInfo = context.packageCache.get(dependencyPackageJsonPath);
+
+    if (packageInfo == null) {
+      const dependencyBaseDir = path.dirname(dependencyPackageJsonPath);
+      const packageJson = readPackageJson(dependencyPackageJsonPath);
+      const depSpecList = objectToDependencySpecList(
+        packageJson.dependencies,
+        packageJson.peerDependencies
+      );
+
+      packageInfo = {
+        version: packageJson.version,
+        source: packageJson._resolved || `local:${fs.realpathSync(dependencyBaseDir)}`,
```

Comment:
Yes, this makes it non-ejectable, we need to fix it. Not sure though how — it can link to a location which is outside the sandbox directory — that way `scp` won't work anyway. We could probably warn in this case.